builtin/logical/database/path_roles.go

```diff
@@ -2,263 +2,623 @@ package database
+	// updating roles, and may use seperate statements depending on the context.
+	createRole := req.Operation == logical.CreateOperation
+	if role == nil {
+		role = &roleEntry{
+			StaticAccount: &staticAccount{},
 		}
+		createRole = true
+	}
 
-		return logical.ListResponse(entries), nil
+	if err := pathRoleCreateUpdateCommon(ctx, role, req.Operation, data); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
 	}
-}
 
-func (b *databaseBackend) pathRoleCreateUpdate() framework.OperationFunc {
-	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-		name := data.Get("name").(string)
-		if name == "" {
-			return logical.ErrorResponse("empty role name attribute given"), nil
+	username := data.Get("username").(string)
+	if username == "" && createRole {
+		return logical.ErrorResponse("username is a required field to create a static account"), nil
+	}
+
+	if role.StaticAccount.Username != "" && role.StaticAccount.Username != username {
+		return logical.ErrorResponse("cannot update static account username"), nil
+	}
+	role.StaticAccount.Username = username
+
+	// If it's a Create operation, both username and rotation_period must be included
+	rotationPeriodSecondsRaw, ok := data.GetOk("rotation_period")
+	if !ok && createRole {
+		return logical.ErrorResponse("rotation_period is required to create static accounts"), nil
+	}
+	if ok {
+		rotationPeriodSeconds := rotationPeriodSecondsRaw.(int)
+		if rotationPeriodSeconds < queueTickSeconds {
+			// If rotation frequency is specified, and this is an update, the value
+			// must be at least that of the constant queueTickSeconds (5 seconds at
+			// time of writing), otherwise we wont be able to rotate in time
+			return logical.ErrorResponse(fmt.Sprintf("rotation_period must be %d seconds or more", queueTickSeconds)), nil
 		}
+		role.StaticAccount.RotationPeriod = time.Duration(rotationPeriodSeconds) * time.Second
+	}
 
-		role, err := b.Role(ctx, req.Storage, data.Get("name").(string))
+	if rotationStmtsRaw, ok := data.GetOk("rotation_statements"); ok {
+		role.Statements.Rotation = rotationStmtsRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		role.Statements.Rotation = data.Get("rotation_statements").([]string)
+	}
+
+	role.StaticAccount.RevokeUserOnDelete = data.Get("revoke_user_on_delete").(bool)
+
+	// lvr represents the roles' LastVaultRotation
+	lvr := role.StaticAccount.LastVaultRotation
+
+	// Only call setStaticAccount if we're creating the role for the
+	// first time
+	switch req.Operation {
+	case logical.CreateOperation:
+		// setStaticAccount calls Storage.Put and saves the role to storage
+		resp, err := b.setStaticAccount(ctx, req.Storage, &setStaticAccountInput{
+			RoleName:   name,
+			Role:       role,
+			CreateUser: createRole,
+		})
 		if err != nil {
 			return nil, err
 		}
-		if role == nil {
-			role = &roleEntry{}
+		// guard against RotationTime not being set or zero-value
```

Comment:
I think it might be preferable to have a zero rotation time if it hasn't happened yet.

Reply:
I would also check what is happening in the AD secrets engine to see what they are doing.

Reply:
I removed the guard. I think it was vestigial, likely the method didn't return an instantiated `time.Time`

Reply:
Oh, you may need to initialize it...maybe.  What does the AD engine do before the first rotation?